Try/new rich text

aerych opened the pull request:
Refs #5594

This PR introduces a new rich text solution in the Reader’s post detail as a replacement to DTCoreText. 

WPRichTextView is replaced in the reader detail by the new WPRichContentView, a subclass of UITextView that relies on Apple’s TextKit framework for rendering HTML in a UITextView. Special treatment is given to blockquote tags, images, and other “embedded” HTML content. 

The new WPRichText classes are based on previous effort in WPTextKit — greatly improved from their WPTextKit incarnation and incorporated directly into WPiOS instead of as a separate library. 

We'll address making the same change in reader comments, removing the DTCoreText pod, and renaming WPRichContentView to WPRichTextView in future PRs.

To test: 
Take the reader detail view for a spin and check for the following things:
- The layout is correct.
- Images are rendered correctly. 
- Images wrapped in a link open the link when tapped
- Images that are not wrapped in a link show the image when tapped.
- Images smaller than the width of the text content are centered.
- Images shorter than the line height are rendered inline.
- Videos (youtube, vimeo, etc.) embedded as iframe are rendered correctly, and play when tapped.
- Confirm that block quotes are indented, italicized, etc.

Confirm that tapping a link opens the in-app browser.

Confirm that posts with lots and lots of images do not suffer from undue lag when loading in the detail.

Confirm that Discover posts have the correct layout for their attribution.


Needs review: @jleandroperez, @koke, could I trouble you both for this one?

cc @frosty @kurzee @bummytime as this is likely relevant to your interests as well.

jleandroperez commented on WordPress/Classes/ViewRelated/Reader/ReaderDetailViewController.swift:
```diff
@@ -497,16 +481,13 @@ public class ReaderDetailViewController : UIViewController, UIViewControllerRest
-        let lineHeight = WPStyleGuide.Detail.contentLineHeight
-        var textOptions = richTextView.textOptions
-        textOptions[DTDefaultFontSize] = WPStyleGuide.Cards.contentFontSize
-        textOptions[DTDefaultLineHeightMultiplier] = lineHeight / fontSize
```
🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥   DT!!

koke replied:
![Kill it with fire](https://cloud.githubusercontent.com/assets/8739/20061666/0de41ace-a501-11e6-94bb-eee67ac127b7.gif)

jleandroperez replied:
@koke i hope you **do** realize.. that gif of yours... is a game changer to me. Right?

koke replied:
I am shocked that this is new to you 😮

jleandroperez replied:
🔥

jleandroperez commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:
```diff
@@ -0,0 +1,267 @@
+        let source = WPTableImageSource(maxSize: self.maxDisplaySize)
+        source.delegate = self
+        source.forceLargerSizeWhenFetching = false
+        source.photonQuality = 65
```
Maybe a constant?

jleandroperez commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:
```diff
@@ -0,0 +1,267 @@
+        }
+        set {
+            let str = newValue ?? ""
+            let style = "<style>body { font-family: Merriweather; font-size:16.0; line-height:1.6875; color: #2e4453; } a { color: #0087be; text-decoration: none; } a:active { color: #005082 } blockquote { font-style: italic; color:#4f748e; } </style>"
```
(Maybe) we could break this in multiple lines, to make it easier to read?

```
let style = "<style>body { font-family: Merriweather; font-size:16.0; line-height:1.6875; color: #2e4453; } " +
"a { color: #0087be; text-decoration: none; } " +
"a:active { color: #005082 } " + 
"blockquote { font-style: italic; color:#4f748e; } </style>"
```

jleandroperez commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:
```diff
@@ -0,0 +1,267 @@
+        _ = attachmentManager
+
+        // Remove some of the unnecessary space at the bottom of the scrollable area.
+        textContainerInset = UIEdgeInsetsMake(0.0, 0.0, -16.0, 0.0)
```
(Maybe) a constant instead?  

```
struct Constants {
   static let defaultContainerInset = ...
}
```

jleandroperez commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:
```diff
@@ -0,0 +1,267 @@
+    ///
+    func embedForAttachment(attachment: WPTextAttachment) -> WPRichTextEmbed {
+        let width: CGFloat = attachment.width ?? textContainer.size.width
+        let height: CGFloat = attachment.height ?? 150.0
```
(Maybe) defaultAttachmentHeight constant?

jleandroperez commented on WordPress/Classes/Networking/ReaderPostServiceRemote.m:
```diff
@@ -893,6 +894,28 @@ - (NSString *)makePlainText:(NSString *)string
 }
 
+
+- (NSString *)removeForbiddenTags:(NSString *)string
```
Can we unit test this fellow?. RegExEs make me dizzy 😄

aerych commented:
Changes made @jleandroperez :)

jleandroperez commented:
Tested with [this post](https://lanteanartest.wordpress.com/2016/11/08/1094/). Few comments sir!

- Youtube isn't starting playback
- Blockquotes seem to always have the same style (italicized without bold)

Other than that, it's looking awesome on a 6s. Will test on an iPad device shortly, **AWESOME** job!!!

jleandroperez commented:
Update: Looks **amazing** on an iPad (+ Readability awesomeness by @kurzee). Same glitches w/ Youtube + Blockquotes, though.

Hopefully easy to fix!!!

frosty commented:
Hey @aerych! Still doing some testing, but I found a post in Discover that reliably crashes `WPRichTextFormatter`: "The Gone Girl With The Dragon Tattoo On The Train" – 3 days ago.

```
2016-11-08 14:53:35.115492 WordPress[5538:2160514] *** Terminating app due to uncaught exception 'NSRangeException', reason: 'NSMutableRLEArray objectAtIndex:effectiveRange:: Out of bounds'
```

App breaks on the `guard` statement of line 97 in `WPRichTextFormatter.swift`.

frosty commented:
A few more things I've noticed whilst testing – sorry for the huge list!

There are a few noticeable differences between the old and new rendering for the recent post "[Inspired by the Cotswolds: A Home Interiors Company that Blogs with Style](https://wordpress.com/read/blogs/53424024/posts/19188)" in Discover.

* There are horizontal rules after the first paragraph and before the last paragraph, which don't show up with the new text view. There's a gap left for them, but they're not visible.
* Blockquotes aren't styled the same (I'm not sure if this is intentional?) – they don't have a coloured background, and the text colour is different.
* If I tap and hold on a link in the old text view, I get a UIMenuController popup with 'copy' as the only action. In the new view, a long press on a link loads the link in the in-app browser (the same as if I'd just tapped on it). It's probably outside the scope of this PR, but personally I'd love it if we could show a `UIActivityViewController` on long press, like you'd get in Safari – allowing the user to do things like share the link to another app.

The post "[Why NaNoWriMo is noble nuttiness – and 8 steps to make it easier](https://wordpress.com/read/blogs/28158232/posts/6855)" in Discover also has a few issues:

* The last paragraph before the first subheading (beginning "Sidenote:" is italicised in the old view, but not in the new one. Same for the 3rd paragraph after the first subheading (After "Oh, no problem, you say:...").
* The indentation for the bullet point list in that section is a little different – but that may be intentional or not an issue.
* In the old view, the padding above and below a bullet point list is the same. In the new view, there is less padding below the list (I think it should probably have more, to match the top). Same for numbered lists.
* Some of the links at the very bottom of the post are also italicised in the old post and not the new one.

Finally, some miscellaneous observations with other test posts:

* Your 'audio test' page on your test blog shows a link to an `m4a` file in the old version, but no link on the new one.
* The latter half of [this post](https://wordpress.com/read/blogs/34823036/posts/5610) (from the table down) just doesn't appear on the new version.
* Your Hamlet test post takes a long time to 'process' before the detail view controller even animates onscreen, and then the text of the post never appears.
* On your Embeds test post, the Soundcloud and Audiomack embeds are the wrong size.
* The Spotify embeds in this post never appear in the new version: https://frostyhappinesstestsite123.com/2015/12/14/title/

 I think what you've done here is amazing! I hope this stuff is useful for you.

koke commented:
This is definitely one of those "I'm trying really hard to break this" bugs, but:

- Visit a reader post with a link
- Tap an hold on the link, and start dragging from the left edge to go back
- Release both fingers
- Tap back and get to an empty view
- Tap back again and crash

https://cloudup.com/cFv0jgAAR0G

aerych commented:
> This is definitely one of those "I'm trying really hard to break this" bugs,

But that's why I like to ping you for reviews :D

aerych commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:
```diff
@@ -0,0 +1,278 @@
+                "body { font-family: Merriweather; font-size:16.0; line-height:1.6875; color: #2e4453; } " +
+                "a { color: #0087be; text-decoration: none; } " +
+                "a:active { color: #005082 } " +
+                "blockquote { font-style: italic; color:#4f748e; } " +
```
I agree but this is probably going to change a bit when we update reader comments.  I'd like to hold off until we see exactly what needs we'll have there.

aerych commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPRichContentView.swift:
```diff
@@ -0,0 +1,278 @@
+                "blockquote { font-style: italic; color:#4f748e; } " +
+                "</style>"
+            let content = style + str
+            let attrTxt = try! NSAttributedString.attributedStringFromHTMLString(content, defaultDocumentAttributes: nil)
```
Paranoia ftw. Updating

aerych commented on WordPress/Classes/ViewRelated/Views/WPRichText/WPTextAttachment.swift:
```diff
@@ -0,0 +1,113 @@
+        } else if height > lineFrag.size.height {
+            // When the attachment is taller than the line height,
+            // reserve the full width for the attachment.  The layout
+            // manager will center the attachment
```
Exactly that.  I'll improve the comment.

aerych commented:
> Youtube isn't starting playback

Thanks for the testing @jleandroperez!  I dug into the issue with the youtube video and it looks like its because some youtube videos are restricted where they can be played. [The video in that post](https://www.youtube.com/embed/ErvgV4P6Fzc?version=3&#038;rel=1&#038;fs=1&#038;autohide=2&#038;showsearch=0&#038;showinfo=1&#038;iv_load_policy=1&#038;wmode=transparent) seems to be one of those.
Here's what happens when I try to play it in safari:
![simulator screen shot nov 10 2016 2 24 26 pm](https://cloud.githubusercontent.com/assets/1435271/20193043/ec9dbdcc-a751-11e6-86e9-65b627a2ccb5.png)

The same issue occurs in the current reader's detail, which is expected really since no changes were made to `WPRichTextEmbed` and its used in both versions of the detail vc. 

I'm inclined to either treat it as a separate issue or to write it off as youtube's bug.  Seems they should display the same error message in a UIWebView that they do in mobile Safari.

Thoughts?

aerych commented:
> Blockquotes seem to always have the same style 

Best i can tell this might be an issue with the Merriweather font as system font's work just fine.  Its easy enough to specify Merriweather-Light in the style tag and that seems to correct the problem.

aerych commented:
> I found a post in Discover that reliably crashes WPRichTextFormatter: "The Gone Girl With The Dragon Tattoo On The Train" – 3 days ago.

Nice find @frosty.  Turns out the post content was an empty string.  Adding a check for this.

aerych commented:
Thanks for the feedback @frosty! 

> * There are horizontal rules after the first paragraph and before the last paragraph, which don't show up with the new text view. There's a gap left for them, but they're not visible.

Horizontal rules are not recognized by NSAttributedString's html parsing. I think we can probably fake them by swapping them out for images, but I want to handle that in a follow up PR.  Thanks for raising this one. It had slipped my mind!

> * Blockquotes aren't styled the same (I'm not sure if this is intentional?) – they don't have a coloured background, and the text colour is different.

This is intentional.  It turns out that when you apply a background color, the background color is not drawn where the first line is indented. I’ve opted for this alternate style for now but I'm hoping we'll be able to sort out the background color eventually.

> * If I tap and hold on a link in the old text view, I get a UIMenuController popup with 'copy' as the only action. In the new view, a long press on a link loads the link in the in-app browser (the same as if I'd just tapped on it). It's probably outside the scope of this PR, but personally I'd love it if we could show a UIActivityViewController on long press, like you'd get in Safari – allowing the user to do things like share the link to another app.

You're wish is my command. :) Added in 8f71bafbf0a89af98b4cd92d44854599c076bbc6

> * The last paragraph before the first subheading (beginning "Sidenote:" is italicised in the old view, but not in the new one. Same for the 3rd paragraph after the first subheading (After "Oh, no problem, you say:...").
    
This was due to a glitch with italics and the Merriweather font.

> * The indentation for the bullet point list in that section is a little different – but that may be intentional or not an issue.

The difference is expected. This is just what we get from the framework.

> * In the old view, the padding above and below a bullet point list is the same. In the new view, there is less padding below the list (I think it should probably have more, to match the top). Same for numbered lists.

Fixed!

> * Some of the links at the very bottom of the post are also italicised in the old post and not the new one.

Same italics glitch as before.  Should be sorted now.

> *    Your 'audio test' page on your test blog shows a link to an m4a file in the old version, but no link on the new one.

A sizing glitch was keeping the audio embed from being visible. Should be fixed now!

> *    The latter half of this post (from the table down) just doesn't appear on the new version.

Not 100% sure of this one.  Seeing this error in the console `NSConcreteTextStorage attribute:atIndex:effectiveRange:]: Range or index out of bounds raised during typesetting layout manager`.  The breakage is where a table appears in the content. The fix for the vanishing text in the Hamlet post might also fix this issue but I need to test more.  Table support is a bit wonky and we've never supported them before so maybe it would be best to just parse them out.  We'll see.

> *    Your Hamlet test post takes a long time to 'process' before the detail view controller even animates onscreen, and then the text of the post never appears.

This was an interesting one. I could reproduce the issue in the app but not in the test harness I had been using when working on the text formatting classes.  The issue is related to how the reader detail's UI is constructed. Currently the textview is a non-scrollable child of a stackview, with the stack view being the content view of a scroll view.  Because the textview is not scrollable, auto layout sizes it to the full height of its content. This worked well enough for the DTCoreText backed WPRichTextView but is causing problems with a UITextView as the text view can no longer draw its text on multiple tiled layers.  It has to draw everything on a single _UITextTiledLayer. If the post is large enough the size of the content exceeds the max _UITextTiledLayer size and the typesetter breaks. I missed this issue in my test harness because its text view *does* scroll.  Anyway, the fix is to restructure the detail vc's UI to preserve the textview's normal scrolling behavior.  Patch coming. :)

> *    On your Embeds test post, the Soundcloud and Audiomack embeds are the wrong size.
> *  The Spotify embeds in this post never appear in the new version: https://frostyhappinesstestsite123.com/2015/12/14/title/

Nice catch. There was an issue with how the height for an attachment was being computed. I found a better way to compute the size of the web content and now we should be good to go.

> I think what you've done here is amazing! I hope this stuff is useful for you.

VERY useful!  Thank you so much!

aerych commented:
Ugh. So, there were some complications.
Regarding this:
> if the post is large enough the size of the content exceeds the max _UITextTiledLayer size and the typesetter breaks. I missed this issue in my test harness because its text view does scroll. Anyway, the fix is to restructure the detail vc's UI to preserve the textview's normal scrolling behavior.

The fix was to use our custom UITextView to scroll all the content and do away with the UIScrollView used previously.  It took some time to find the best way to correctly position the post header and attribution view as subviews of a UITextView, and how best to preserve readable content margins.  The approach I chose relies on setting text container insets to create space for the header and footer, and preserve the left and right margins.  However, it seems that setting a top inset has the side effect of increasing the content offset when changing orientation.  Its something that can be reproduced 10/10 in the simplest project containing just a UITextView, an NSAttributedString large enough to scroll, and setting the top textContainerInset to, say 50px or higher.
I'm not happy with the scroll offset increasing when changing orientation, but I'm not sure it should be a blocker for the feature, and I've delayed this patch longer than I should while looking for a fix.  That said I'm probably going to tinker with it a bit more while this last batch of changes is in review.

Apart from the changes to the detail vc's view hierarchy I've made a couple of fixes to the formatter and added tables to our forbidden tags.  They were never supported in the past and they are proving to be problematic here. I guess they won't be missed. 

All that said... I think this is ready for another round.  @frosty @jleandroperez @koke thanks for being patient.

aerych commented:
I've tinkered a bit with an [alternative means](https://github.com/wordpress-mobile/WordPress-iOS/commit/d204b2ec1df8b2e6a06740e5e75f5bc5cdde38ea) of making space in the textView for the header and discover attribution.  You can see it in this [alternate branch](https://github.com/wordpress-mobile/WordPress-iOS/tree/try/new-rich-text-alternate).  The approach forgoes setting insets and instead relies on empty NSTextAttachments to reserve space.  I think its a bit wonky but it seems to avoid or greatly reduce issues with the content offset when changing orientation.  This is most obvious with text only posts.  Posts with images that scale on rotation still throw off the scroll offset and I'm starting to think this is unavoidable.

Anyway, because its kind of a weird approach I wanted to solicit opinions before merging it here.

frosty commented:
Awesome work, @aerych! I've retested things I mentioned before, and your fixes look great. The only slightly 'odd' thing is that italic text and blockquotes look a little strange to me, although I can't really put my finger on why:

![simulator screen shot 22 nov 2016 14 16 17](https://cloud.githubusercontent.com/assets/4780/20528420/08c21c8e-b0c4-11e6-8c11-5e63ff95e9dc.png)
![simulator screen shot 22 nov 2016 14 14 23](https://cloud.githubusercontent.com/assets/4780/20528422/08eb14ea-b0c4-11e6-89d8-ef4941de5d2b.png)

I _think_ it almost feels like it's a different font? If nothing else, I wonder if the line spacing is a little wide on the blockquotes, too – should it match the main body text?

aerych commented:
> I think it almost feels like it's a different font?

They are actually two separate fonts.  One is Merriweather, the other is Merriweather-Italic.  For some reason Merriweather does not get assigned an italic trait when the html is converted to an NSAttributedString.  My hunch is there is something missing in the font file that the system is looking for.  Maybe we can drop Merriweather in favor of the system font since it works as expected.

> I wonder if the line spacing is a little wide on the blockquotes

Did you mean too much space between the lines?  In the example above, there are two separate paragraphs vs one paragraph with multiple lines. 

Curious what you though of https://github.com/wordpress-mobile/WordPress-iOS/pull/6170#issuecomment-261822263 since this is something we chatted about the other week.

frosty commented:
> Did you mean too much space between the lines? In the example above, there are two separate paragraphs vs one paragraph with multiple lines.

<facepalm.gif>

Regarding https://github.com/wordpress-mobile/WordPress-iOS/pull/6170#issuecomment-261822263, I think if you're happier with the result with text-only posts (I agree it does seem to help a lot with posts with resizing images, but that's not the end of the world), then it looks like a reasonable way to do things :)

aerych commented:
> looks like a reasonable way to do things

It still feels odd to me.  Maybe its "oddly reasonable"?  "Reasonably odd"?  One of those? :)

Thanks for the gut check on the approach. I do think its worth it for the text-only posts so I'll get it merged in.

jleandroperez commented:
Small finding: Blockquotes seem to always be italicized:

This snippet:

```
Block Quotes:
<blockquote>This is a Quote</blockquote>
Block Quotes Bold:
<blockquote><strong>This is a Bold Quote</strong></blockquote>
Block Quotes Italicized
<blockquote><em>This is an Italic Quote</em></blockquote>
Block Quotes Bold  Italics:
<blockquote><em><strong>This is an Italicized Bold Quote</strong></em></blockquote>
```

Translates into:

![img_5791](https://cloud.githubusercontent.com/assets/1195260/20537747/06169c0e-b0cd-11e6-8dd8-359af2fe1d42.PNG)

jleandroperez commented:
RE: Youtube not playing, we should probably give some sort of feedback, but i'm totally fine with addressing that (+ the blockquotes thing) in another issue.

aerych commented:
> Blockquotes seem to always be italicized

This is intentional, along with the different text color.

aerych commented:
> RE: Youtube not playing, we should probably give some sort of feedback

I'm not sure we can tell whether the video plays or not.  We don't control the content and wouldn't even know about the restriction unless (perhaps) there was an error when it loaded. 

Interestingly, something's changed because I'm now seeing the Youtube error message.  Not sure if its due to changes made in the branch or something Youtube is doing differently.

![simulator screen shot nov 22 2016 1 09 57 pm](https://cloud.githubusercontent.com/assets/1435271/20538092/5000b7e0-b0b5-11e6-9ec2-3059c09bba51.png)

jleandroperez commented:
### iPad + Rotations:

1. Open [this testing post](on an iPad device)
2. Rotate the device
3. As a result, the text might end up overlapping with the images

![simulator screen shot nov 22 2016 16 14 34](https://cloud.githubusercontent.com/assets/1195260/20538200/ed669ba8-b0ce-11e6-9a3e-ab95a5054094.png)

jleandroperez commented:
iPad + Rotations fix confirmed.

Let's go ahead and merge this puppy, and deal with (whatever new bug arise) in their own Issue / PR. The sooner we get a beta with TextKit, the better.

:shipit: on my end sir, awesome job!!!

**P.s.:** Congratz on removing DT!!!!!

aerych commented:
> Congratz on removing DT!!!!!

We still have to do the reader comments before its official. Can't wait!

aerych commented:
Thanks @jleandroperez !!  I think we've addressed all the concerns, including a couple Koke pinged me about on Slack.  Shipping with your squirrel and we can address any other issues as we find them in the beta. :)
